alg/los.cpp

```diff
@@ -0,0 +1,224 @@
+
+	}
+	else
+	{
+		dx <<= 1;
+		balance = dx - dy;
+		dy <<= 1;
+
+		while (y != y2  && isAboveTerrain)
+		{
+			isAboveTerrain &= OnBresenhamPoint(x, y);
+			if (balance >= 0)
+			{
+				x += incx;
+				balance -= dy;
+			}
+			balance += dx;
+			y += incy;
+		}
+        isAboveTerrain &= OnBresenhamPoint(x, y);
+	}
+    return isAboveTerrain;
+}
+
+
+// Get the elevation of a single point.
+bool GetElevation(const GDALRasterBandH hBand, const int x, const int y, double& val)
+{
+    CPLAssert(x >= 0);
+    CPLAssert(x <= GDALGetRasterBandXSize(hBand));
+    CPLAssert(y >= 0);
+    CPLAssert(y <= GDALGetRasterBandYSize(hBand));
+
+    return GDALRasterIO(hBand, GF_Read, x, y, 1, 1, &val, 1, 1, GDT_Float64, 0, 0) == CE_None;
+}
+
+// Check a single location is above terrain.
+bool IsAboveTerrain(const GDALRasterBandH hBand, const int x, const int y, const double z)
+{
+    double terrainHeight;
+    if(GetElevation(hBand, x, y, terrainHeight)) {
+        return z > terrainHeight;
+    } else {
+        return false;
+    }
+}
+
+/************************************************************************/
+/*                        GDALIsLineOfSightVisible()                    */
+/************************************************************************/
+
+/**
+ * Check Line of Sight between two points.
+ * Both input coordinates must be within the raster coordinate bounds.
+ *
+ * This algorithm will check line of sight using a 3D bresenham algorithm.
+ * https://www.researchgate.net/publication/2411280_Efficient_Line-of-Sight_Algorithms_for_Real_Terrain_Data
+ *
+ * @param hBand The band to read the DEM data from.
+ * 
+ * @param xA The X location to check on the raster to read the DEM data from.
+ *
+ * @return True if the two points are within Line of Sight.
+ *
+ * @since GDAL 3.X
+ */
+
+bool GDALIsLineOfSightVisible(
+    const GDALRasterBandH hBand, const int xA, const int yA, const double zA, const int xB, const int yB, const double zB, const char** papszOptions)
+{
+    if(hBand == nullptr) {
+        return false;
+    }
+
+    // Perform a preliminary check of the start and end points.
+    if (!IsAboveTerrain(hBand, xA, yA, zA)) {
+        return false;
+    
+    }
+    if (!IsAboveTerrain(hBand, xB, yB, zB)) {
+        return false;
+    }
+
+    // If both X and Y are the same, no further checks are needed.
+    if(xA == xB && yA == yB) {
+        return true;
+    }
+
+    // TODO if both X's or Y's are the same, it could be optimized for vertical/horizontal lines.
+
+
+    // Use an interpolated Z height with 2D bresenham for the remaining cases.
+
+    // Lambda for Linear interpolate like C++20 std::lerp.
+    auto lerp = [](const int a, const int b, const float t) {
+        return a + t * (b - a);
+    };
+
+    // Lambda for getting Z test height given x input along the bresenham line.
+    auto GetZValue = [&](const int x) -> double {
+        const auto xDiff = xB - xA;
+        const auto xPercent = x / xDiff;
```

Comment:
xB and xA are int, so xDiff is int.  x is int as well, so x / xDiff is an integer division, which doesn't seem to be what you want. There should be a cast of one member to float, so presumably:
```suggestion
        const auto xPercent = float(x) / xDiff;
```

Reply:
yea, I'll fix this. When it was all double math earlier, it was fine.

Reply:
Resolved.